File: Repo/opencog/cogserver/Julia/cogserver.jl
#!/usr/bin/env julia
#=
cogserver.jl

OpenCog CogServer - Julia Implementation
Network server and REPL interface for OpenCog

This implementation demonstrates Julia's strengths:
- TCP networking with concurrent client handling
- Task-based asynchronous programming
- Higher-order functions for command handlers
- Module system with closures
- Dictionary-based command registry
=#

using Sockets
using Dates
using Printf

# ============================================================================
# COMMAND SYSTEM
# ============================================================================

"""
    CommandHandler

Type alias for command handler functions.
Takes arguments vector, returns (result_string, success_bool)
"""
const CommandHandler = Function

"""
    Command

Represents a registered command in the CogServer.
Demonstrates: Struct with function fields, documentation
"""
struct Command
    name::String
    description::String
    handler::CommandHandler
end

# ============================================================================
# MODULE INTERFACE
# ============================================================================

"""
    CogModule

Abstract type for CogServer modules.
Modules can register commands and manage their own state.
"""
abstract type CogModule end

# Module interface (to be implemented by concrete modules)
module_init(m::CogModule, server) = error("module_init not implemented")
module_commands(m::CogModule) = error("module_commands not implemented")
module_shutdown(m::CogModule) = error("module_shutdown not implemented")

# ============================================================================
# COGSERVER IMPLEMENTATION
# ============================================================================

"""
    CogServer

Main server structure handling network connections and command execution.
Demonstrates: Mutable struct, concurrent programming, event loop
"""
mutable struct CogServer
    address::String
    port::Int
    server_socket::Union{TCPServer, Nothing}
    commands::Dict{String, Command}
    modules::Dict{String, CogModule}
    running::Bool
    client_tasks::Vector{Task}
    lock::ReentrantLock
    
    function CogServer(address::String="0.0.0.0", port::Int=17001)
        cs = new(
            address,
            port,
            nothing,
            Dict{String, Command}(),
            Dict{String, CogModule}(),
            false,
            Task[],
            ReentrantLock()
        )
        
        # Register built-in commands
        register_builtin_commands!(cs)
        
        return cs
    end
end

# ============================================================================
# COMMAND REGISTRATION
# ============================================================================

"""
    register_command!(cs::CogServer, cmd::Command)

Register a new command in the server.
"""
function register_command!(cs::CogServer, cmd::Command)
    lock(cs.lock) do
        cs.commands[cmd.name] = cmd
    end
end

"""
    unregister_command!(cs::CogServer, name::String)

Unregister a command from the server.
"""
function unregister_command!(cs::CogServer, name::String)
    lock(cs.lock) do
        delete!(cs.commands, name)
    end
end

# ============================================================================
# BUILT-IN COMMAND HANDLERS
# ============================================================================

function handle_help(cs::CogServer, args::Vector{String})
    lock(cs.lock) do
        if isempty(cs.commands)
            return ("No commands available", true)
        end
        
        result = "Available commands:\n"
        for (name, cmd) in sort(collect(cs.commands), by=x->x[1])
            result *= @sprintf("  %-15s - %s\n", name, cmd.description)
        end
        return (result, true)
    end
end

function handle_list(cs::CogServer, args::Vector{String})
    lock(cs.lock) do
        if isempty(cs.modules)
            return ("No modules loaded", true)
        end
        
        result = "Loaded modules:\n"
        for (name, _) in sort(collect(cs.modules), by=x->x[1])
            result *= "  - $name\n"
        end
        return (result, true)
    end
end

function handle_shutdown(cs::CogServer, args::Vector{String})
    println("[CogServer] Shutdown command received")
    cs.running = false
    return ("Shutting down server...", true)
end

function handle_info(cs::CogServer, args::Vector{String})
    lock(cs.lock) do
        num_commands = length(cs.commands)
        num_modules = length(cs.modules)
        num_clients = length(cs.client_tasks)
        
        result = """
        CogServer Information:
          Address: $(cs.address):$(cs.port)
          Running: $(cs.running)
          Commands: $num_commands
          Modules: $num_modules
          Active Clients: $num_clients
        """
        return (result, true)
    end
end

function handle_echo(cs::CogServer, args::Vector{String})
    if isempty(args)
        return ("Usage: echo <message>", false)
    end
    return (join(args, " "), true)
end

"""
    register_builtin_commands!(cs::CogServer)

Register all built-in commands.
Demonstrates: Higher-order functions, closures
"""
function register_builtin_commands!(cs::CogServer)
    # Help command
    register_command!(cs, Command(
        "help",
        "Display available commands",
        args -> handle_help(cs, args)
    ))
    
    # List modules command
    register_command!(cs, Command(
        "list",
        "List loaded modules",
        args -> handle_list(cs, args)
    ))
    
    # Shutdown command
    register_command!(cs, Command(
        "shutdown",
        "Shutdown the server",
        args -> handle_shutdown(cs, args)
    ))
    
    # Info command
    register_command!(cs, Command(
        "info",
        "Display server information",
        args -> handle_info(cs, args)
    ))
    
    # Echo command
    register_command!(cs, Command(
        "echo",
        "Echo back arguments",
        args -> handle_echo(cs, args)
    ))
end

# ============================================================================
# MODULE MANAGEMENT
# ============================================================================

"""
    load_module!(cs::CogServer, name::String, module::CogModule)

Load a module and register its commands.
"""
function load_module!(cs::CogServer, name::String, module::CogModule)
    lock(cs.lock) do
        if haskey(cs.modules, name)
            return (false, "Module $name already loaded")
        end
        
        # Initialize module
        try
            module_init(module, cs)
        catch e
            return (false, "Failed to initialize module $name: $e")
        end
        
        # Register module commands
        for cmd in module_commands(module)
            cs.commands[cmd.name] = cmd
        end
        
        cs.modules[name] = module
        return (true, "Module $name loaded successfully")
    end
end

"""
    unload_module!(cs::CogServer, name::String)

Unload a module and unregister its commands.
"""
function unload_module!(cs::CogServer, name::String)
    lock(cs.lock) do
        if !haskey(cs.modules, name)
            return (false, "Module $name not loaded")
        end
        
        module = cs.modules[name]
        
        # Unregister module commands
        for cmd in module_commands(module)
            delete!(cs.commands, cmd.name)
        end
        
        # Shutdown module
        try
            module_shutdown(module)
        catch e
            return (false, "Failed to shutdown module $name: $e")
        end
        
        delete!(cs.modules, name)
        return (true, "Module $name unloaded successfully")
    end
end

# ============================================================================
# SERVER OPERATIONS
# ============================================================================

"""
    start!(cs::CogServer)

Start the CogServer and begin accepting connections.
Demonstrates: Asynchronous programming, task management
"""
function start!(cs::CogServer)
    lock(cs.lock) do
        if cs.running
            error("Server already running")
        end
        
        # Create TCP server
        try
            cs.server_socket = listen(IPv4(cs.address), cs.port)
        catch e
            error("Failed to start listener: $e")
        end
        
        cs.running = true
        println("[CogServer] Listening on $(cs.address):$(cs.port)")
    end
    
    # Accept connections in main loop
    while cs.running
        try
            # Accept with timeout to allow checking running flag
            client_sock = accept(cs.server_socket)
            
            # Handle client in separate task
            task = @async handle_client(cs, client_sock)
            push!(cs.client_tasks, task)
            
            # Clean up finished tasks
            filter!(t -> !istaskdone(t), cs.client_tasks)
        catch e
            if cs.running
                println("[CogServer] Error accepting connection: $e")
            end
        end
    end
    
    # Cleanup
    cleanup!(cs)
end

"""
    stop!(cs::CogServer)

Stop the CogServer.
"""
function stop!(cs::CogServer)
    println("[CogServer] Stopping server...")
    cs.running = false
    
    if cs.server_socket !== nothing
        close(cs.server_socket)
    end
end

"""
    cleanup!(cs::CogServer)

Clean up server resources.
"""
function cleanup!(cs::CogServer)
    println("[CogServer] Cleaning up...")
    
    # Wait for client tasks to finish (with timeout)
    deadline = time() + 5.0
    while !isempty(cs.client_tasks) && time() < deadline
        filter!(t -> !istaskdone(t), cs.client_tasks)
        sleep(0.1)
    end
    
    # Close server socket
    if cs.server_socket !== nothing
        close(cs.server_socket)
        cs.server_socket = nothing
    end
    
    println("[CogServer] Cleanup complete")
end

# ============================================================================
# CLIENT HANDLING
# ============================================================================

"""
    handle_client(cs::CogServer, client_sock::TCPSocket)

Handle a connected client.
Demonstrates: REPL implementation, I/O handling
"""
function handle_client(cs::CogServer, client_sock::TCPSocket)
    try
        # Send welcome message
        println(client_sock, "Welcome to CogServer (Julia)")
        println(client_sock, "Type 'help' for available commands")
        print(client_sock, "cogserver> ")
        flush(client_sock)
        
        # Read-eval-print loop
        while cs.running && isopen(client_sock)
            line = ""
            try
                line = readline(client_sock)
            catch e
                if e isa EOFError
                    break
                end
                println("[CogServer] Error reading from client: $e")
                break
            end
            
            line = strip(line)
            
            if isempty(line)
                print(client_sock, "cogserver> ")
                flush(client_sock)
                continue
            end
            
            # Execute command
            response = execute_command(cs, line)
            println(client_sock, response)
            
            if !cs.running
                println(client_sock, "Server shutting down...")
                break
            end
            
            print(client_sock, "cogserver> ")
            flush(client_sock)
        end
    catch e
        println("[CogServer] Client handler error: $e")
    finally
        close(client_sock)
    end
end

"""
    execute_command(cs::CogServer, line::String) -> String

Execute a command line and return the result.
"""
function execute_command(cs::CogServer, line::String)
    parts = split(line)
    if isempty(parts)
        return ""
    end
    
    cmd_name = String(parts[1])
    args = String[String(p) for p in parts[2:end]]
    
    # Look up command
    cmd = lock(cs.lock) do
        get(cs.commands, cmd_name, nothing)
    end
    
    if cmd === nothing
        return "Unknown command: $cmd_name"
    end
    
    # Execute command handler
    try
        result, success = cmd.handler(args)
        return result
    catch e
        return "Error executing command: $e"
    end
end

# ============================================================================
# DEMONSTRATION FUNCTION
# ============================================================================

"""
    demo()

Demonstrate CogServer functionality with interactive mode.
"""
function demo()
    println("="^70)
    println("OpenCog CogServer - Julia Implementation Demo")
    println("Network server and REPL interface")
    println("="^70)
    println()
    
    println("1. Creating CogServer")
    println("-"^70)
    server = CogServer("127.0.0.1", 17001)
    println("  Server created on 127.0.0.1:17001")
    println("  Built-in commands: $(length(server.commands))")
    println()
    
    println("2. Available Commands")
    println("-"^70)
    result, _ = handle_help(server, String[])
    println(result)
    
    println("3. Server Information")
    println("-"^70)
    result, _ = handle_info(server, String[])
    println(result)
    
    println("4. Testing Commands Locally")
    println("-"^70)
    test_commands = ["echo Hello OpenCog!", "info", "list"]
    for cmd in test_commands
        println("  > $cmd")
        response = execute_command(server, cmd)
        println(response)
    end
    println()
    
    println("5. Starting Server (Press Ctrl+C to stop)")
    println("-"^70)
    println("Connect with: telnet 127.0.0.1 17001")
    println("Or: nc 127.0.0.1 17001")
    println()
    
    # Start server in demo mode
    println("Starting server...")
    println("Press Ctrl+C to stop the server")
    println()
    
    try
        start!(server)
    catch e
        if e isa InterruptException
            println("\nInterrupt received, stopping server...")
            stop!(server)
        else
            rethrow(e)
        end
    end
    
    println()
    println("="^70)
    println("Julia CogServer strengths demonstrated:")
    println("  ✓ TCP networking with Sockets module")
    println("  ✓ Asynchronous client handling with tasks")
    println("  ✓ Higher-order functions for command handlers")
    println("  ✓ Module system with dynamic loading")
    println("  ✓ REPL-style interactive interface")
    println("  ✓ Thread-safe operations with locks")
    println("="^70)
end

# Run demo if executed as script
if abspath(PROGRAM_FILE) == @__FILE__
    demo()
end

File: Repo/opencog/cogserver/Go/cogserver.go
// Package cogserver provides a Go implementation of the OpenCog CogServer
// network server and REPL interface.
//
// The CogServer provides network access to an AtomSpace and supports
// command execution, module loading, and agent scheduling.
package cogserver

import (
	"bufio"
	"fmt"
	"io"
	"net"
	"strings"
	"sync"
)

// CommandHandler is a function that processes a command
type CommandHandler func(args []string) (string, error)

// Command represents a registered command
type Command struct {
	Name        string
	Description string
	Handler     CommandHandler
}

// Module represents a loadable module
type Module interface {
	Init(server *CogServer) error
	GetCommands() []Command
	Shutdown() error
}

// CogServer is the main server structure
type CogServer struct {
	address  string
	listener net.Listener
	commands map[string]Command
	modules  map[string]Module
	running  bool
	mu       sync.RWMutex
	wg       sync.WaitGroup
}

// NewCogServer creates a new CogServer
func NewCogServer(address string) *CogServer {
	server := &CogServer{
		address:  address,
		commands: make(map[string]Command),
		modules:  make(map[string]Module),
	}
	
	// Register built-in commands
	server.registerBuiltinCommands()
	
	return server
}

// registerBuiltinCommands registers the default commands
func (cs *CogServer) registerBuiltinCommands() {
	cs.RegisterCommand(Command{
		Name:        "help",
		Description: "Display available commands",
		Handler:     cs.handleHelp,
	})
	
	cs.RegisterCommand(Command{
		Name:        "list",
		Description: "List loaded modules",
		Handler:     cs.handleList,
	})
	
	cs.RegisterCommand(Command{
		Name:        "shutdown",
		Description: "Shutdown the server",
		Handler:     cs.handleShutdown,
	})
}

// RegisterCommand registers a new command
func (cs *CogServer) RegisterCommand(cmd Command) {
	cs.mu.Lock()
	defer cs.mu.Unlock()
	cs.commands[cmd.Name] = cmd
}

// UnregisterCommand removes a command
func (cs *CogServer) UnregisterCommand(name string) {
	cs.mu.Lock()
	defer cs.mu.Unlock()
	delete(cs.commands, name)
}

// LoadModule loads a module and registers its commands
func (cs *CogServer) LoadModule(name string, module Module) error {
	cs.mu.Lock()
	defer cs.mu.Unlock()
	
	if _, exists := cs.modules[name]; exists {
		return fmt.Errorf("module %s already loaded", name)
	}
	
	if err := module.Init(cs); err != nil {
		return fmt.Errorf("failed to initialize module %s: %w", name, err)
	}
	
	// Register module commands
	for _, cmd := range module.GetCommands() {
		cs.commands[cmd.Name] = cmd
	}
	
	cs.modules[name] = module
	return nil
}

// UnloadModule unloads a module and unregisters its commands
func (cs *CogServer) UnloadModule(name string) error {
	cs.mu.Lock()
	defer cs.mu.Unlock()
	
	module, exists := cs.modules[name]
	if !exists {
		return fmt.Errorf("module %s not loaded", name)
	}
	
	// Unregister module commands
	for _, cmd := range module.GetCommands() {
		delete(cs.commands, cmd.Name)
	}
	
	if err := module.Shutdown(); err != nil {
		return fmt.Errorf("failed to shutdown module %s: %w", name, err)
	}
	
	delete(cs.modules, name)
	return nil
}

// Start starts the server
func (cs *CogServer) Start() error {
	cs.mu.Lock()
	if cs.running {
		cs.mu.Unlock()
		return fmt.Errorf("server already running")
	}
	
	listener, err := net.Listen("tcp", cs.address)
	if err != nil {
		cs.mu.Unlock()
		return fmt.Errorf("failed to start listener: %w", err)
	}
	
	cs.listener = listener
	cs.running = true
	cs.mu.Unlock()
	
	fmt.Printf("CogServer listening on %s\n", cs.address)
	
	// Accept connections
	cs.wg.Add(1)
	go cs.acceptConnections()
	
	return nil
}

// acceptConnections accepts incoming connections
func (cs *CogServer) acceptConnections() {
	defer cs.wg.Done()
	
	for {
		conn, err := cs.listener.Accept()
		if err != nil {
			cs.mu.RLock()
			running := cs.running
			cs.mu.RUnlock()
			
			if !running {
				return
			}
			fmt.Printf("Error accepting connection: %v\n", err)
			continue
		}
		
		cs.wg.Add(1)
		go cs.handleConnection(conn)
	}
}

// handleConnection handles a client connection
func (cs *CogServer) handleConnection(conn net.Conn) {
	defer cs.wg.Done()
	defer conn.Close()
	
	fmt.Fprintf(conn, "Welcome to CogServer\n")
	fmt.Fprintf(conn, "Type 'help' for available commands\n")
	fmt.Fprintf(conn, "cogserver> ")
	
	scanner := bufio.NewScanner(conn)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		
		if line == "" {
			fmt.Fprintf(conn, "cogserver> ")
			continue
		}
		
		response := cs.executeCommand(line)
		fmt.Fprintf(conn, "%s\n", response)
		
		// Check if server is shutting down
		cs.mu.RLock()
		running := cs.running
		cs.mu.RUnlock()
		
		if !running {
			fmt.Fprintf(conn, "Server shutting down...\n")
			return
		}
		
		fmt.Fprintf(conn, "cogserver> ")
	}
	
	if err := scanner.Err(); err != nil && err != io.EOF {
		fmt.Printf("Connection error: %v\n", err)
	}
}

// executeCommand executes a command string
func (cs *CogServer) executeCommand(line string) string {
	parts := strings.Fields(line)
	if len(parts) == 0 {
		return ""
	}
	
	cmdName := parts[0]
	args := parts[1:]
	
	cs.mu.RLock()
	cmd, exists := cs.commands[cmdName]
	cs.mu.RUnlock()
	
	if !exists {
		return fmt.Sprintf("Unknown command: %s", cmdName)
	}
	
	result, err := cmd.Handler(args)
	if err != nil {
		return fmt.Sprintf("Error: %v", err)
	}
	
	return result
}

// Stop stops the server
func (cs *CogServer) Stop() error {
	cs.mu.Lock()
	if !cs.running {
		cs.mu.Unlock()
		return fmt.Errorf("server not running")
	}
	
	cs.running = false
	
	// Close listener
	if cs.listener != nil {
		cs.listener.Close()
	}
	cs.mu.Unlock()
	
	// Wait for all connections to close
	cs.wg.Wait()
	
	// Unload all modules
	cs.mu.Lock()
	for name := range cs.modules {
		cs.modules[name].Shutdown()
	}
	cs.modules = make(map[string]Module)
	cs.mu.Unlock()
	
	fmt.Println("CogServer stopped")
	return nil
}

// Built-in command handlers

func (cs *CogServer) handleHelp(args []string) (string, error) {
	cs.mu.RLock()
	defer cs.mu.RUnlock()
	
	var sb strings.Builder
	sb.WriteString("Available commands:\n")
	
	for name, cmd := range cs.commands {
		sb.WriteString(fmt.Sprintf("  %-15s %s\n", name, cmd.Description))
	}
	
	return sb.String(), nil
}

func (cs *CogServer) handleList(args []string) (string, error) {
	cs.mu.RLock()
	defer cs.mu.RUnlock()
	
	if len(cs.modules) == 0 {
		return "No modules loaded", nil
	}
	
	var sb strings.Builder
	sb.WriteString("Loaded modules:\n")
	
	for name := range cs.modules {
		sb.WriteString(fmt.Sprintf("  %s\n", name))
	}
	
	return sb.String(), nil
}

func (cs *CogServer) handleShutdown(args []string) (string, error) {
	go func() {
		cs.Stop()
	}()
	return "Shutting down server...", nil
}

// IsRunning returns whether the server is running
func (cs *CogServer) IsRunning() bool {
	cs.mu.RLock()
	defer cs.mu.RUnlock()
	return cs.running
}

// GetAddress returns the server address
func (cs *CogServer) GetAddress() string {
	return cs.address
}
